Download updates

LennartPurucker opened the pull request:
Related to #1034 #1081 #1132 
Closes #1081 #1132 

In detail, in relation to #1034, I have added a deprecation warning for `get_dataset`, making clear that we will default to not downloading the data in the future (starting from version 1.4, but this is just my guess for now). 

In relation to #1081, I added an option to disable downloading features. Moreover, I adjusted the `repr` method and added semi-lazy loading for features by the use of the function `OpenMLDataset.load_metadata`. As part of this, I also added the option to load qualities after calling `get_dataset`, which seems to have been missing so far. 

In relation to #1132, I added the option `force_refresh_cache` to force refresh the cache by deleting the existing cache initially. 

Note, I have also changed the docstring of `get_dataset` and adjusted the parts about being threading/multiprocessing safe, which were incorrect from my previous experiences and observations w.r.t. the cache. This might open a discussion to make this threading/multiprocessing safe based on an option that only downloads the data and ignores the cache entirely. I advise against this, for now, see [here ](https://github.com/openml/openml-python/pull/1256/files#diff-9cc0e191426810b9a76098707c2ddc9ec73a7f1852024a72445ddcb08b52f09bR361-R367) for more details on this.

LennartPurucker commented:
@mfeurer I have not found a good way to put these functionalities into a general function among all OpenML entity types. Hence, I would like your opinion on the current proposed way of doing it. Does it suffice? 
Feel free to remark on additional reference points (entities) that would need such download-related treatment so that I can better understand how to make these changes more abstract.

PGijsbers commented on openml/datasets/data_feature.py:
```diff
@@ -62,5 +62,17 @@ def __init__(
+    def __eq__(self, other):
+        if not isinstance(other, OpenMLDataFeature):
+            return False
+        else:
```
You can drop the `else` and dedent the remainder.

PGijsbers commented on openml/datasets/dataset.py:
```diff
@@ -773,6 +788,31 @@ def get_data(
         return data, targets, categorical, attribute_names
 
+    def load_metadata(self, features: bool = False, qualities: bool = False):
+        """Load the missing medata information from the server and store it in the server.
```
```suggestion
        """Load the missing metadata information from the server and store it in the server.
```

PGijsbers commented on openml/datasets/dataset.py:
```diff
@@ -773,6 +788,31 @@ def get_data(
+            )
+
+        if features and self.features is None:
+            self.features = _parse_features_xml(_get_features_xml(self.dataset_id))
```
Prefer one function call per line, it is much more readable if an (unexpected) error is raised.

PGijsbers reviewed (commented):
didn't have a close look, had a very small amount of time, still needs a review (not necessarily by me).

mfeurer commented on openml/datasets/data_feature.py:
```diff
@@ -62,5 +62,17 @@ def __init__(
+        if not isinstance(other, OpenMLDataFeature):
+            return False
+
+        return (
```
Could we simplify this code by comparing `self.__dict__` with `other.__dict__`?

mfeurer commented on openml/datasets/functions.py:
```diff
@@ -383,21 +386,42 @@ def get_dataset(
         Option to download 'qualities' meta-data in addition to the minimal dataset description.
+        If True, download and cache the qualities file.
+        If False, create the OpenMLDataset without qualities metadata. The data may later be added
+        to the OpenMLDataset hrough the `OpenMLDataset.load_metadata(qualities=True)` method.
```
```suggestion
        to the OpenMLDataset through the `OpenMLDataset.load_metadata(qualities=True)` method.
```

mfeurer commented on openml/datasets/functions.py:
```diff
@@ -355,13 +352,19 @@ def get_dataset(
+    A check will be performed to determine if the information has previously been downloaded to a
+    cache, and if so be loaded from disk instead of retrieved from the server.
+
+    To make this function thread/multiprocessing safe initialize the cache first by calling
```
You could also extend the documentation that this becomes thread-safe when oslo.concurrency is installed (see the decorator).

mfeurer commented on openml/datasets/functions.py:
```diff
@@ -383,21 +386,42 @@ def get_dataset(
         The downloaded dataset.
     """
+    if any([download_qualities, download_features_meta_data]):
+        warnings.warn(
```
I would suggest approaching this a bit differently:
1. We set the flag to `None` above and emit a `DeprecationWarning` then.
2. We ask the user to pass a `bool` to disable this warning.
3. We internally assume `True` if `None` is given.
4. And also do this for `download_dataset`.

LennartPurucker replied:
Your idea sounds much better. I will change this accordingly.

One question, I noticed that Python, by default, suppresses deprecation warnings. Should we maybe switch to user warnings? Otherwise, users only see the warning if they enable deprecation warnings. 

We could also change the warning level in our code, but this would change settings in the outer scope unless we clean up properly.

LennartPurucker replied:
I added your suggestion and switched to user warnings for now.

mfeurer replied:
scikit-learn appears to use `FutureWarning`, so we might use that instead?

mfeurer commented on openml/datasets/functions.py:
```diff
@@ -419,6 +443,15 @@ def get_dataset(
 
+    # Note: we could also (quite heavily) re-implement the below to only download the
+    # data and do not cache the data at all. This would always be thread/multiprocessing
+    # safe. However, this would likely drastically increase the strain on the server.
```
And make any work with OpenML really slow. Please see my comment above on how to make this class thread-safe.

PGijsbers replied:
+1, we definitely want to cache data. In the vast amount of use cases it is more important than making this function thread safe.

mfeurer commented on tests/test_datasets/test_dataset.py:
```diff
@@ -262,6 +262,19 @@ def test_get_data_corrupt_pickle(self):
+        )
+
+        _dataset = openml.datasets.get_dataset(
+            2, download_data=False, download_features_meta_data=False, download_qualities=False
```
Hm, interesting. So the data is already available on the local drive, so you would not download it again, so the argument name is wrong. This should actually be called `load_features_meta_data` then, I guess?

LennartPurucker replied:
Not sure what you mean exactly. Calling `get_dataset(download_features_meta_data=True)` will download features meta_data or load it from the cache. This works similarly for `download_data` and `download_qualities`. Technically, all of these are a bit misleading, but considering that it is a cache I would say it is okay. 

If you referring to the function call below, `OpenMLDataset.load_metadata(features=True)`, the code will always download the data. I think I should also add a cache for that as `load_data` also caches the data or loads from the cache if available. I will change this accordingly.

LennartPurucker replied:
I have added cache usage to `OpenMLDataset.load_metadata`. It is now in line with `get_dataset()` and `OpenMLDataset.get_data`'s `_load_data`.

mfeurer replied:
Nevermind, I'm not sure whether my comment makes too much sense.

LennartPurucker replied:
It was a good comment. I have changed a few things to adjust to your comment and I the code is now better/clearer.

mfeurer reviewed (commented):
Hey, I left some comments. All in all this looks great and I'm optimistic that we can merge this soonish.

Just a side note: we'd also need to update the task loading to disable dataset loading by default.

LennartPurucker commented:
I have also adjusted task loading. This got a bit hacky so that it is backward compatible and supports all arguments by passing kwargs. I am not sure that this is a good solution. Hence,  I would be happy for feedback on how to handle the arguments best. 

Moreover, I added actually lazy loading based on properties such that the user does not have to call `load_metadata` by hand. This is now more in line with how `get_data` functions as far as I understand it.

mfeurer commented on openml/runs/functions.py:
```diff
@@ -102,8 +102,8 @@ def run_model_on_task(
-            "https://openml.github.io/openml-python/main/examples/20_basic/introduction_tutorial.html#authentication"
-            "for more information on authentication.",
+            "https://openml.github.io/openml-python/main/examples/20_basic/introduction_tutorial"
+            + ".html#authenticatio for more information on authentication.",
```
Did you drop an `n` here?

mfeurer commented on openml/tasks/functions.py:
```diff
@@ -327,27 +326,39 @@ def get_tasks(
-    downloaded optionally based on the additional parameter. Else,
-    splits will either way be downloaded when the task is being used.
+    Downloads the task representation, while the data splits can be downloaded optionally based on
+    the additional parameter that are passed to :meth:`openml.datasets.get_dataset`.
```
I don't quite understand this. If the `download_splits` is given, this has nothing to do with the additional parameter, or does it?

LennartPurucker replied:
I will update the doc string to make it more clear.

mfeurer commented on openml/datasets/dataset.py:
```diff
@@ -831,10 +853,6 @@ def retrieve_class_labels(self, target_name: str = "class") -> Union[None, List[
         -------
         list
         """
-        if self.features is None:
```
Why can this be dropped?

LennartPurucker replied:
To clarify, I dropped this because with the new lazy loading,` self.feature` cannot be `None`.
AFAIK, it would make no sense to call this function with the intention of `self.feature is None`, because the function requires values for `self.feature`. Hence, I think defaulting to lazy loading is appropriate here.

mfeurer commented on openml/tasks/functions.py:
```diff
@@ -327,31 +326,54 @@ def get_tasks(
+        download_splits = get_dataset_kwargs.get("download_data", True)
+
     if not isinstance(task_id, int):
+        # TODO(0.14): Remove warning
```
```suggestion
        # TODO(0.15): Remove warning
```

LennartPurucker replied:
Done!

mfeurer reviewed (changes requested):
1 more change, then we can merge this one!

PGijsbers commented on openml/datasets/dataset.py:
```diff
@@ -234,12 +239,36 @@ def find_invalid_characters(string, pattern):
 
+    @property
+    def features(self):
+        # Lazy loading of features
```
I think this is a common enough pattern to not warrant a comment.

PGijsbers commented on openml/datasets/dataset.py:
```diff
@@ -922,6 +981,7 @@ def _to_dict(self) -> "OrderedDict[str, OrderedDict]":
         return data_container
 
 
+# -- Code for Features Property
```
I'd rather open a new issue on Github as a note that we need to refactor our files later. Having code headers shouldn't really be necessary.

PGijsbers commented on openml/tasks/functions.py:
```diff
@@ -327,31 +326,54 @@ def get_tasks(
-    downloaded optionally based on the additional parameter. Else,
-    splits will either way be downloaded when the task is being used.
+    Downloads the task representation. By default, this will also download the data splits and
+    the dataset. From version 0.15.0 onwards, the splits will not be downloaded by default
+    nor the dataset.
```
```suggestion
    the dataset. From version 0.15.0 onwards, the splits nor the dataset will not be downloaded by default.
```

PGijsbers commented on tests/test_datasets/test_dataset.py:
```diff
@@ -262,6 +262,36 @@ def test_get_data_corrupt_pickle(self):
+
+        _dataset = openml.datasets.get_dataset(
+            2, download_data=False, download_features_meta_data=False, download_qualities=False
+        )
```
Should check cache is empty after this call? A function which always loads the data should now pass this test.

PGijsbers commented on tests/test_datasets/test_dataset_functions.py:
```diff
@@ -546,8 +546,47 @@ def test__get_dataset_qualities(self):
+            did_cache_dir,
+        )
+
+        # Test clean start
```
This section should be a separate test.

PGijsbers reviewed (changes requested):
I didn't look too much at the correctness of the code (since Matthias already had a look), but I think there are several things that could be improved from a code quality point of view. Please have a look at the comments.

LennartPurucker commented on openml/datasets/dataset.py:
```diff
@@ -773,6 +802,40 @@ def get_data(
+
+        if qualities_file is not None:
+            self._qualities = _read_qualities(qualities_file)
+            self._no_qualities_found = self._qualities is None
```
This is required as far as I can tell, because `self._no_qualities_found` is initially False and only becomes True in this edge case after lazy loading the qualities for the first time.
